docs/design/2022-09-29-reorganize-partition.md

```diff
@@ -0,0 +1,169 @@
+
+The operation should be online, and must handle multiple partitions as well as large data sets.
+
+Possible usage scenarios:
+- Full table copy
+  - merging all partitions to a single table (ALTER TABLE t REMOVE PARTITIONING)
+  - splitting data from many to many partitions, like change the number of hash partitions
+  - splitting a table to many partitions (ALTER TABLE t PARTITION BY ...)
+- Partial table copy (not full table/all partitions)
+  - split one or more partitions
+  - merge two or more partitions
+
+These different use cases can have different optimizations, but the generic form must still be solved:
+- N partitions, where each partition has M indexes
+
+First implementation should be based on the merge-txn (row-by-row read, update record, write) transactional batches.
+Later we can implement the ingest (lightning way) optimization, since it is not yet completely done.
+
+## Detailed Design
+
+There are two parts of the design:
+- Schema change states throughout the operation
+- Reorganization implementation
+
+Where the schema change states will clarify which different steps that will be done in which schema state transisions.
+
+### Schema change states for REORGANIZE PARTITION
+
+Since this operation will:
+- create new partitions
+- drop existing partitions
+- copy data from dropped partitions to new partitions
+- 
+it will use all these schema change stages:
+
+      // StateNone means this schema element is absent and can't be used.
+      StateNone SchemaState = iota
+      - Check if the table structure after the ALTER is valid
+      - Generate physical table ids to each new partition
+      - Update the meta data with the new partitions and which partitions to be dropped (so that new transaction can double write)
+      - TODO: Should we also set placement rules? (Lable Rules)
+      - Set the state to StateDeleteOnly
+      
+      // StateDeleteOnly means we can only delete items for this schema element.
+      StateDeleteOnly
+      - Set the state to StateWriteOnly
+
+      // StateWriteOnly means we can use any write operation on this schema element,
+      // but outer can't read the changed data.
+      StateWriteOnly
+      - Set the state to StateWriteReorganization
+      
+      // StateWriteReorganization means we are re-organizing whole data after write only state.
+      StateWriteReorganization
+      - Copy the data from the partitions to be dropped and insert it into the new partitions [1]
+      - Write the index data for the new partitions [2]
+      - Replace the old partitions with the new partitions in the metadata
+      - Add a job for dropping the old partitions data (which will also remove its placement rules etc.)
```

Comment:
Here we can drop the old partitions data with regist old partition id in delet ranges when  this job finishing.

Reply:
What I mean is when we finished job and change the partitions to be public state, the DDL task flow will provide a function to register old partition id into job and finishDDLJob will check if this kind of DDL task need GC data and do the delete data task Async.
![2osgoe5mXr](https://user-images.githubusercontent.com/13519188/195068249-2df6d55c-b8c8-430d-89b8-d6be6f3784d2.png)

So you do not need to start another job to delete old partition data.

Reply:
Thank you for the clarification, I have updated the design accordingly.